sql: refactor USING constraint planning

sploiselle opened the pull request:
Our previous approach to planning `USING` constraints wasn't totally sufficient because it precluded filtering _NULL_ values from the output due to its reliance on `coalesce` over the joined columns (which explicitly removes _NULL_ values in the face of other, non-_NULL_ values).

This refactor instead:
- Leaves all columns from the joined relation intact, and instead sets some column in the joined relation as the "priority" version of the column
- Modifies `SELECT *` to automatically unproject any non-priority version of priority columns

Because prioritization is/was previously influenced/dictated by the column's presence in the projection list, this shouldn't interact with the other places in the code using prioritization...even though maybe it should!

@benesch pinging you only because you implemented name prioritization in the first place, so maybe you have feedback on how this approach is subtly wrong?

### `coalesce`

The previous implementation relied on coalescing columns in all cases, and that approach would have worked here. However, by only creating new, coalesced, prioritized columns when strictly necessary (on `FullOuterJoin`), and prioritizing existing columns in other cases, we get some nice wins in the optimizer.

### Motivation

This PR fixes a recognized bug:
Fixes #7618

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

benesch commented on src/sql/src/plan/query.rs:
```diff
@@ -2488,16 +2488,33 @@ fn expand_select_item<'a>(
+                    }
                 })
                 .collect();
+
```
Oh man, this scope stuff just gets messier and messier. This logic makes me fairly nervous. Can we instead add a `visible_to_wildcard: bool` to `ScopeItem` that's frobbed correctly by the outer join logic and just filter on that bool here?

sploiselle replied:
Yep, not a problem.

benesch reviewed (approved):
Code LGTM! DIdn't think too hard about whether this was a sufficient set of tests, since I figured we have @philip-stoev for that. :)

benesch commented:
Wait, sorry, we want the `visible_to_wildcard` on the `ScopeItem`, not the `ScopeItemName`!

```diff
diff --git a/src/sql/src/plan/query.rs b/src/sql/src/plan/query.rs
index 66f0df983..9b6b49948 100644
--- a/src/sql/src/plan/query.rs
+++ b/src/sql/src/plan/query.rs
@@ -1635,6 +1635,7 @@ fn plan_view_select(
                         names: vec![],
                         expr: group_expr.cloned(),
                         nameable: true,
+                        visible_to_wildcard: true,
                     }
                 };
 
@@ -1660,6 +1661,7 @@ fn plan_view_select(
                 names: vec![],
                 expr: Some(Expr::Function(sql_function.clone())),
                 nameable: true,
+                visible_to_wildcard: true,
             });
         }
         if !agg_exprs.is_empty() || !group_key.is_empty() || having.is_some() {
@@ -1725,7 +1727,6 @@ fn plan_view_select(
                         table_name: None,
                         column_name: Some(column_name.clone()),
                         priority: true,
-                        visible_to_wildcard: true,
                     });
                 }
             } else {
@@ -1738,11 +1739,11 @@ fn plan_view_select(
                             table_name: None,
                             column_name: Some(column_name.clone()),
                             priority: true,
-                            visible_to_wildcard: true,
                         })
                         .collect(),
                     expr: None,
                     nameable: true,
+                    visible_to_wildcard: true,
                 });
             }
         }
@@ -2350,7 +2351,6 @@ fn plan_table_function_alias(
                 table_name: None,
                 column_name: Some(normalize::column_name(alias.name.clone())),
                 priority: false,
-                visible_to_wildcard: true,
             });
         }
     }
@@ -2387,7 +2387,6 @@ fn plan_table_alias(mut scope: Scope, alias: Option<&TableAlias>) -> Result<Scop
                 }),
                 column_name,
                 priority: false,
-                visible_to_wildcard: true,
             }];
         }
     }
@@ -2429,7 +2428,6 @@ fn invent_column_name(ecx: &ExprContext, expr: &Expr<Aug>) -> Option<ScopeItemNa
         table_name: None,
         column_name: Some(n),
         priority: false,
-        visible_to_wildcard: true,
     })
 }
 
@@ -2498,12 +2496,10 @@ fn expand_select_item<'a>(
                 .items
                 .iter()
                 .enumerate()
-                .filter_map(|(i, item)| match item.names.get(0) {
-                    Some(n) if !n.visible_to_wildcard => None,
-                    name => {
-                        let name = name.and_then(|n| n.column_name.clone());
-                        Some((ExpandedSelectItem::InputOrdinal(i), name))
-                    }
+                .filter(|(_i, item)| item.visible_to_wildcard)
+                .map(|(i, item)| {
+                    let name = item.names.get(0).and_then(|n| n.column_name.clone());
+                    (ExpandedSelectItem::InputOrdinal(i), name)
                 })
                 .collect();
 
@@ -2789,7 +2785,6 @@ fn plan_using_constraint(
                         table_name: None,
                         column_name,
                         priority: true,
-                        visible_to_wildcard: true,
                     })
                     .collect::<Vec<_>>();
 
@@ -2802,6 +2797,7 @@ fn plan_using_constraint(
                     names,
                     expr: None,
                     nameable: true,
+                    visible_to_wildcard: true,
                 });
             }
         }
@@ -2818,15 +2814,15 @@ fn plan_using_constraint(
     for c in &non_priority_cols {
         for item_name in both_scope.items[*c].names.iter_mut() {
             item_name.priority = false;
-            item_name.visible_to_wildcard = false;
         }
+        both_scope.items[*c].visible_to_wildcard = false;
     }
 
     for c in &priority_cols {
         for item_name in both_scope.items[*c].names.iter_mut() {
             item_name.priority = true;
-            item_name.visible_to_wildcard = true;
         }
+        both_scope.items[*c].visible_to_wildcard = true;
     }
 
     // Reproject all returned elements to the front of the list.
diff --git a/src/sql/src/plan/scope.rs b/src/sql/src/plan/scope.rs
index 5b90fd2cc..41dee8a75 100644
--- a/src/sql/src/plan/scope.rs
+++ b/src/sql/src/plan/scope.rs
@@ -74,7 +74,8 @@ pub struct ScopeItemName {
     ///
     /// For example, in the following `SELECT` statement
     ///
-    ///   CREATE TABLE t (a int, b) SELECT 'outer' AS a, b FROM t ORDER BY a
+    ///   CREATE TABLE t (a int, b)
+    ///   SELECT 'outer' AS a, b FROM t ORDER BY a
     ///
     /// even though there are two columns named `a` in scope in the ORDER BY
     /// (one from `t` and one declared in the select list), the column declared
@@ -85,12 +86,6 @@ pub struct ScopeItemName {
     ///
     /// this still needs to generate an "ambiguous name" error.
     pub priority: bool,
-    /// Controls whether or not the item should return from `SELECT *`.
-    ///
-    /// It's possible that this feature could be folded into `priority`, but it
-    /// isn't clear what the long-term ramifications are with SQL support in
-    /// that case.
-    pub visible_to_wildcard: bool,
 }
 
 #[derive(Debug, Clone)]
@@ -107,6 +102,12 @@ pub struct ScopeItem {
     // column names. Omitting the name entirely is not an option, since the name
     // is used to label the column in the result set.
     pub nameable: bool,
+    /// Controls whether or not the item should return from `SELECT *`.
+    ///
+    /// It's possible that this feature could be folded into `priority`, but it
+    /// isn't clear what the long-term ramifications are with SQL support in
+    /// that case.
+    pub visible_to_wildcard: bool,
 }
 
 #[derive(Debug, Clone)]
@@ -124,10 +125,10 @@ impl ScopeItem {
                 table_name: None,
                 column_name,
                 priority: false,
-                visible_to_wildcard: true,
             }],
             expr: None,
             nameable: true,
+            visible_to_wildcard: true,
         }
     }
 
@@ -185,10 +186,10 @@ impl Scope {
                     table_name: table_name.clone(),
                     column_name: column_name.map(|n| n.into()),
                     priority: false,
-                    visible_to_wildcard: true,
                 }],
                 expr: None,
                 nameable: true,
+                visible_to_wildcard: true,
             })
             .collect();
         scope
```

sploiselle commented:
@benesch Ah--that seems much more sensible. Ty for catching that/thinking of it.

sploiselle commented on test/sqllogictest/joins.slt:
```diff
@@ -692,3 +691,259 @@ SELECT lb, rb FROM l3 INNER JOIN r3 ON (la IS NULL AND ra IS NULL) OR la = ra
+CREATE TABLE IF NOT EXISTS t4 (f3 int, f4 int);
+
+statement ok
+DELETE FROM t1;
```
Sure thing!

philip-stoev commented:
Apologies for the delay.

Item No#1. This query fails in both the prior commit and in Postgres, but succeeds in this branch:

```
SELECT *
  FROM t1 AS a1
  JOIN t1 AS a2 USING (f1, f2)
  JOIN t1 AS a3 USING (f1)
  JOIN t1 AS a4 USING (f2);
ERROR:  common column name "f2" appears more than once in left table
```

philip-stoev commented:
Item No #2. This query fails in this branch and the prior commit, but succeeds in postgres. Not sure if this is within scope of this PR:

```
materialize=> SELECT * FROM t2 AS a1 , t2 AS a3  JOIN t1 AS a4 USING ( f2 );
ERROR:  column name "f2" is ambiguous
```

DDL:

```
CREATE TABLE t1 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
CREATE TABLE t2 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
```

philip-stoev commented:
Item No #3. This query:

```
SELECT a3.f1,
       a4.f1
  FROM t2 AS a1
  JOIN t1 AS a2 USING (f1) , t2 AS a3
  JOIN t1 AS a4 USING (f1);
```

Is interpreted differently as follows:
- in the prior commit -- error column name "f1" is ambiguous
- in this commit, 12 rows , that is, the last USING is interpreted as `ON (a1.f1 = a4.f1)`
- in Postgres, 4 rows, that is, the last USING is interpreted as `ON (a3.f1 = a4.f1)`
- MySQL - 4 rows
- SQLite - 12 rows.


DDL:

```
CREATE TABLE t1 (f1 INTEGER, f2 INTEGER NOT NULL);

INSERT INTO t1 VALUES (NULL, 0);

INSERT INTO t1 VALUES (1, 1);
INSERT INTO t1 VALUES (1, 1);
INSERT INTO t1 VALUES (2, 2);

CREATE TABLE t2 (f1 INTEGER, f2 INTEGER NOT NULL);
INSERT INTO t2 VALUES (NULL, 0);
INSERT INTO t2 VALUES (NULL, 0);
INSERT INTO t2 VALUES (1, 1);
```

philip-stoev commented:
Item No#4. Postgres does not allow duplicate column names in the `USING` clause, but this branch and the prior commit do. I do not know why Postgres has this restriction, I am unable to figure out a situation where this would lead to a wrong result:

```
pstoev=# select * from t1 join t2 using (f1,f1);
ERROR:  column name "f1" appears more than once in USING clause
```

This check was introduced as a part of a larger 20-year-old Postgres commit and does not provide any rationale for the restriction. Several other databases appears to have inherited the same error message.

philip-stoev reviewed (approved):
Thank you for the tests! I am able to see this fix in action:
- the column counts are now correct with respect to Postgres
- certain wrong results with subqueries not involving explicit IS NULL are now fixed

So the only thing that remains is to decide upon the queries mentioned as individual comments-- they are all of the "3 joins per query" variety.

sploiselle commented:
@philip-stoev

No 1. Fixed; we were not clearing the priority on creating a new scope

No 2, 3. Fixed. I believe the same issue caused both of these, which is fixed in the last commit on this PR. In essence, comma-joined tables should behave as nested cross joins. This also fixed #6875. Given that this was unexpected, would you mind taking a look at this PR again?

No 4. Fixed with obvious approach. Doesn't seem like it hurts, but also doesn't seem to matter much?

benesch commented:
Color me massively confused on the #6875 fix. It seemed like #6875 was massively hard to fix and it... wasn't?? Anyway, an AST transform is ... well, this seems like maybe the wrong moment to use an AST transform. Can we fix either the parser or the planner instead?

sploiselle commented:
@benesch I had the same code in the planner and it works, no sweat. My thinking here was that the "comma" is really syntactic sugar that unrolls into something more complicated. However, if that line of reasoning isn't compelling, certainly glad to plop it down elsewhere.

benesch commented:
> @benesch I had the same code in the planner and it works, no sweat. My thinking here was that the "comma" is really syntactic sugar that unrolls into something more complicated. However, if that line of reasoning isn't compelling, certainly glad to plop it down elsewhere.

Ah, sorry, my point is that there is a bug in `plan_table_with_joins`, and it should just be fixed in `plan_table_with_joins` rather than with an AST rewrite. Here's what I mean!

```diff
diff --git a/src/sql/src/plan/query.rs b/src/sql/src/plan/query.rs
index 101338f9a..4cb27c52f 100644
--- a/src/sql/src/plan/query.rs
+++ b/src/sql/src/plan/query.rs
@@ -36,8 +36,8 @@ use sql_parser::ast::fold::Fold;
 use sql_parser::ast::visit::{self, Visit};
 use sql_parser::ast::{
     Assignment, AstInfo, Cte, DataType, DeleteStatement, Distinct, Expr, Function, FunctionArgs,
-    Ident, InsertSource, IsExprConstruct, Join, JoinConstraint, JoinOperator, Limit, OrderByExpr,
-    Query, Raw, RawName, Select, SelectItem, SetExpr, SetOperator, Statement, SubscriptPosition,
+    Ident, InsertSource, IsExprConstruct, JoinConstraint, JoinOperator, Limit, OrderByExpr, Query,
+    Raw, RawName, Select, SelectItem, SetExpr, SetOperator, Statement, SubscriptPosition,
     TableAlias, TableFactor, TableFunction, TableWithJoins, UnresolvedObjectName, UpdateStatement,
     Value, Values,
 };
@@ -1531,8 +1531,6 @@ fn plan_view_select(
     };
 
     // Step 1. Handle FROM clause, including joins.
-    let from = rewrite_select_from(from.clone());
-
     let (mut relation_expr, from_scope) =
         from.iter().fold(Ok(plan_join_identity(qcx)), |l, twj| {
             let (left, left_scope) = l?;
@@ -1877,80 +1875,6 @@ fn plan_view_select(
     })
 }
 
-/// Comma-joined tables should be rewritten as nested cross joins to
-/// properly handle join precedence. This effect is perhaps most noticeable
-/// when interspersing `NATURAL`/`USING`-constrained joins and comma joins,
-/// e.g. the schema of...
-/// ```sql
-/// CREATE TABLE t1 (f1 int, f2 int);
-/// CREATE TABLE t2 (f1 int, f2 int);
-/// SELECT *
-///      FROM t2,
-///      t2 AS x
-///      JOIN t1
-///      USING (f2);
-/// ```
-///
-/// is very confounding until you realize it's equivalent to the query...
-///
-/// ```sql
-/// SELECT *
-///      FROM t2
-///      CROSS JOIN (
-///      t2 AS x
-///          JOIN t1
-///          USING (f2)
-///      );
-/// ```
-///
-/// i.e. the `USING` constraint applies to `x` and `t1`, which is then
-/// cross-joined with `t2`.
-fn rewrite_select_from(mut twjs: Vec<TableWithJoins<Aug>>) -> Vec<TableWithJoins<Aug>> {
-    let mut nested_join_candidate = None;
-
-    // The new `from` value will get filled in backward and reversed before
-    // placing it back in the AST.
-    let mut from: Vec<TableWithJoins<Aug>> = Vec::with_capacity(twjs.len());
-
-    // Note this process is done R to L.
-    for mut twj in twjs.drain(..).rev() {
-        let mut lateral_visitor = LateralJoinVisitor::new();
-        lateral_visitor.visit_table_with_joins(&twj);
-
-        // Nested joins are only applicable for adjacently listed tables that do
-        // not contain lateral joins.
-        if matches!(twj.relation, TableFactor::Table { .. }) && !lateral_visitor.found() {
-            nested_join_candidate.map(|njc| {
-                // Place the last-seen table in this table's joins as a
-                // nested cross join.
-                twj.joins.push(Join {
-                    relation: TableFactor::NestedJoin {
-                        join: Box::new(njc),
-                        alias: None,
-                    },
-                    join_operator: JoinOperator::CrossJoin,
-                })
-            });
-            // This table is now eligible to be cross-joined with its
-            // predecessor.
-            nested_join_candidate = Some(twj);
-        } else {
-            // Place `nested_join_candidate` back into `from` because it is
-            // not cross-joined with any non-table objects.
-            nested_join_candidate.take().map(|njc| from.push(njc));
-            from.push(twj);
-        }
-    }
-
-    // If the last element we looked at was a table, we need to place it
-    // back at the "beginning" of the list.
-    nested_join_candidate.take().map(|njc| from.push(njc));
-
-    // We looked at everything R to L, so reverse it.
-    from.reverse();
-    from
-}
-
 /// Plans an expression in a `GROUP BY` clause.
 ///
 /// For historical reasons, PostgreSQL allows `GROUP BY` expressions to refer to
@@ -2075,29 +1999,61 @@ fn plan_table_with_joins<'a>(
     left_scope: Scope,
     table_with_joins: &'a TableWithJoins<Aug>,
 ) -> Result<(HirRelationExpr, Scope), PlanError> {
-    let pushed_left_was_join_identity = left.is_join_identity();
-    let (mut left, mut left_scope) = plan_table_factor(
-        qcx,
-        left,
-        left_scope,
-        &JoinOperator::CrossJoin,
-        &table_with_joins.relation,
-    )?;
-    for join in &table_with_joins.joins {
-        if !pushed_left_was_join_identity
-            && matches!(
-                &join.join_operator,
-                JoinOperator::FullOuter(..) | JoinOperator::RightOuter(..)
-            )
-        {
-            bail_unsupported!(6875, "full/right outer joins in comma join");
+    let mut lateral_visitor = LateralJoinVisitor::new();
+    lateral_visitor.visit_table_with_joins(&table_with_joins);
+    if !lateral_visitor.found() {
+        // NOTE: this branch builds the correct join tree, but it bricks LATERAL
+        // behavior. See #6875.
+        let (expr, scope) = plan_join_identity(qcx);
+        let (mut expr, mut scope) = plan_table_factor(
+            qcx,
+            expr,
+            scope,
+            &JoinOperator::CrossJoin,
+            &table_with_joins.relation,
+        )?;
+        for join in &table_with_joins.joins {
+            let (new_expr, new_scope) =
+                plan_table_factor(qcx, expr, scope, &join.join_operator, &join.relation)?;
+            expr = new_expr;
+            scope = new_scope;
         }
-        let (new_left, new_left_scope) =
-            plan_table_factor(qcx, left, left_scope, &join.join_operator, &join.relation)?;
-        left = new_left;
-        left_scope = new_left_scope;
+        plan_join_operator(
+            &JoinOperator::CrossJoin,
+            qcx,
+            left,
+            left_scope,
+            &qcx,
+            expr,
+            scope,
+            false,
+        )
+    } else {
+        // This branch is WRONG, but it mostly handles LATERAL correctly.
+        let pushed_left_was_join_identity = left.is_join_identity();
+        let (mut left, mut left_scope) = plan_table_factor(
+            qcx,
+            left,
+            left_scope,
+            &JoinOperator::CrossJoin,
+            &table_with_joins.relation,
+        )?;
+        for join in &table_with_joins.joins {
+            if !pushed_left_was_join_identity
+                && matches!(
+                    &join.join_operator,
+                    JoinOperator::FullOuter(..) | JoinOperator::RightOuter(..)
+                )
+            {
+                bail_unsupported!(6875, "full/right outer joins in comma join");
+            }
+            let (new_left, new_left_scope) =
+                plan_table_factor(qcx, left, left_scope, &join.join_operator, &join.relation)?;
+            left = new_left;
+            left_scope = new_left_scope;
+        }
+        Ok((left, left_scope))
     }
-    Ok((left, left_scope))
 }
 
 fn plan_table_factor(
@@ -2654,19 +2610,19 @@ fn plan_join_operator(
             // column references in `right`, but it doesn't seem worth it just
             // to make the raw query plans nicer. The optimizer doesn't have
             // trouble with the extra join.)
-            let join = if left.is_join_identity() && !lateral {
-                right
+            if left.is_join_identity() && !lateral {
+                Ok((right, right_scope))
             } else if right.is_join_identity() {
-                left
+                Ok((left, left_scope))
             } else {
-                HirRelationExpr::Join {
+                let join = HirRelationExpr::Join {
                     left: Box::new(left),
                     right: Box::new(right),
                     on: HirScalarExpr::literal_true(),
                     kind: JoinKind::Inner { lateral },
-                }
-            };
-            Ok((join, left_scope.product(right_scope)?))
+                };
+                Ok((join, left_scope.product(right_scope)?))
+            }
         }
     }
 }
@@ -4212,10 +4168,11 @@ impl LateralJoinVisitor {
 
 impl<'a, 'ast> Visit<'ast, Aug> for LateralJoinVisitor {
     fn visit_table_factor(&mut self, node: &'ast TableFactor<Aug>) {
-        if let TableFactor::Derived { lateral, .. } = node {
-            if *lateral {
-                self.found = true;
-            }
+        if matches!(
+            node,
+            TableFactor::Derived { lateral: true, .. } | TableFactor::Function { .. }
+        ) {
+            self.found = true;
         }
     }
 }
```

That said I find this absolutely terrifying. I'm really nervous about having different join trees generated based on whether or not there are `LATERAL` elements present. I think we should investigate solving the scoping issues described in #6875 for real.

asenac commented:
AFAICS #6875 isn't fully fixed in this branch. It makes us support more cases but there are some we don't support yet:

```
materialize=> explain raw plan for select * from t1, t2 full join lateral(select t1.f1 from t3) on true;
ERROR:  full/right outer joins in comma join not yet supported, see https://github.com/MaterializeInc/materialize/issues/6875 for more details
materialize=>
```

philip-stoev commented:
@sploiselle no problem, please let me know when the dust has settled and I will retest.

sploiselle commented:
@asenac If I remove the check for the query you provided, I am now able to catch another, more-general and well known error (we only support `LATERAL` references for `INNER` and `LEFT` joins). I am not suggesting this is definitively the correct approach, but do you have any other suggestions for queries that might be likely to fail given this change? To make it easy to poke around with, I've pushed the changes I'm working with into this PR.

sploiselle commented:
@benesch tysm for the patch

asenac commented:
As discussed, the proposed solution looks good to me. The new error for the query in my previous comment is more concrete after removing the old one, which allows us to support more cases not involving lateral joins like this one:

```
materialize=> explain raw plan for select * from t1, t2 full join t3 on true;
             Raw Plan             
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %1 =                            +
 | Get materialize.public.t2 (u7)+
                                 +
 %2 =                            +
 | Get materialize.public.t3 (u9)+
                                 +
 %3 =                            +
 | FullOuterJoin %1 %2 on true   +
                                 +
 %4 =                            +
 | InnerJoin %0 %3 on true       +
 
(1 row)

materialize=>
```

benesch commented:
Ok, I've managed to construct a query that exhibits wrong results due to the issue described in #6875 _if_ the [check for invalid joins](https://github.com/MaterializeInc/materialize/blob/60ae96c1364aee10fa7903e080cc4d1b6ea0c158/src/sql/src/plan/query.rs#L2042-L2049) is removed:

```sql
CREATE TABLE t1 (a int, b int);
CREATE TABLE t2 (a int, c int);
INSERT INTO t1 VALUES (1, 2), (2, 3);
INSERT INTO t2 VALUES (2, 4), (5, 7);
SELECT * FROM generate_series(1, 2), LATERAL (SELECT * FROM t1) _ NATURAL RIGHT JOIN t2;
```

@sploiselle, do you mind whipping that up into an SLT? I'll feel much better knowing that we've got at least one test that will fail if that check is removed.

Can we also change the error message from "full/right outer joins in comma join not yet supported" to "join trees that mix lateral joins and full/right outer joins not yet supported", or somesuch?

benesch reviewed (approved):
This looks wonderful. Thanks so much, @sploiselle!

sploiselle commented:
@philip-stoev Oh derp--I got over-eager and merged this because it seemed like the right fix. If you have time, I would very much welcome additional testing and will prioritize fixing any issues you uncover.

philip-stoev commented:
@sploiselle no problem. Further PRs are expected in the general vicinity so I will return to USING in due course.
